### PR TITLE
[QC-549] Add repocleaner production rule

### PR DESCRIPTION
**Pull request by Barthelemy**

### CODE REVIEW

File: Framework/script/RepoCleaner/rules/production.py [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+
+        # run is finished for long enough
+        if eor is not None and datetime.now() > eor + timedelta(minutes=delay_final_trimming):
+            logging.debug("      Run is over for long enough, let's do the final trimming")
+            final_trimming(ccdb, period_btw_versions_final, run_versions, preservation_list, update_list, deletion_list)
+
+        else:  # trim the versions as the run is ongoing or too fresh
+            logging.debug("      Run is too fresh or still ongoing, we do the light trimming")
+            first_trimming(ccdb, delay_first_trimming, period_btw_versions_first, run_versions,
+                           preservation_list, update_list, deletion_list)
```
Comment (knopers8) [inline Framework/script/RepoCleaner/rules/production.py:117]: will this actually behave correctly with an empty line between if and else?

File: Framework/script/RepoCleaner/rules/production.py [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+    for v in run_versions:
+        logging.debug(f"      Processing {v} ")
+        # if it is the first or if the last_preserved is older than `period_btw_versions_final`
+        if last_preserved is None \
+                or last_preserved.validFromAsDt < v.validFromAsDt - timedelta(minutes=period_btw_versions_final) \
+                or v == run_versions[-1]:  # v is last element, which we must preserve
+            if last_preserved is not None:  # first extend validity of the previous preserved and set flag
+                ccdb.updateValidity(last_preserved, last_preserved.validFrom, str(int(v.validFrom) - 1), metadata)
+                update_list.append(last_preserved)
+                logging.debug(f"      Extention of {last_preserved}")
```
Comment (knopers8) [inline Framework/script/RepoCleaner/rules/production.py:180]: Extension

**Review by knopers8 — CHANGES_REQUESTED** [review]
Thanks, I don't see any major problems, I just left a few pedantic comments.

File: Framework/script/RepoCleaner/Ccdb.py [hunk 3]
```diff
@@ -114,31 +113,68 @@ def deleteVersion(self, version: ObjectVersion):
...
+    @dryable.Dryable()
+    def setMetadata(self, key: str, value: str):
+        '''
+
+        :param key:
+        :param value:
+        :return:
+        '''
+        logging.debug(f"Set key {key} to {value}")
+
```
[the remark this answers is not in the document]
Comment (knopers8) [reply Framework/script/RepoCleaner/Ccdb.py:150]: Ah, it's because it is for the dry run, isn't it?

[the remark this answers is not in the document]
Comment (Barthelemy) [reply Framework/script/RepoCleaner/Ccdb.py:150]: It is, I will add a comment.

File: Framework/script/RepoCleaner/Ccdb.py [hunk 4]
```diff
@@ -15,27 +15,26 @@ class ObjectVersion:
...
         :param path: path to the object
         :param uuid: unique id of the object
-        :param validFromAsDatetime: validity range smaller limit (in ms)
+        :param validFrom: validity range smaller limit (in ms)
         :param validTo: validity range bigger limit (in ms)
         '''
         self.path = path
         self.uuid = uuid
-        self.validFromAsDatetime = datetime.datetime.fromtimestamp(validFrom / 1000)  # /1000 because we get ms 
+        self.validFromAsDt = datetime.datetime.fromtimestamp(int(validFrom) / 1000)  # /1000 because we get ms
```
[the remark this answers is not in the document]
Comment (Barthelemy) [reply Framework/script/RepoCleaner/Ccdb.py:28]: done

File: Framework/script/RepoCleaner/Ccdb.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Barthelemy) [reply Framework/script/RepoCleaner/Ccdb.py:150]: Actually it was not used, I removed it.

File: Framework/script/RepoCleaner/rules/production.py [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+from collections import defaultdict
+from datetime import datetime
+import time
+from datetime import timedelta, date
+
+from Ccdb import Ccdb, ObjectVersion
+
+
+def in_grace_period(version: ObjectVersion, delay: int):
+    return not (version.validFromAsDt < datetime.now() - timedelta(minutes=delay))
```
[the remark this answers is not in the document]
Comment (Barthelemy) [reply Framework/script/RepoCleaner/rules/production.py:12]: agreed

File: Framework/script/RepoCleaner/rules/production.py (hunk 1 above)
Comment (Barthelemy) [reply Framework/script/RepoCleaner/rules/production.py:117]: yes, it does but you are correct that it does not respect the usual coding standard of python. Fixed.

**Review by knopers8 — APPROVED** [review]

**Pull request merged**